tests/config-array.test.js

```diff
@@ -1291,6 +1291,240 @@ describe('ConfigArray', () => {
+				configs = new ConfigArray([
+					{
+						files: ['**/*.js']
+					},
+					{
+						ignores: [
+							'foo/**',
+
+							// only the directory is unignored, files are not
+							'!foo'
+						]
+					}
+				], {
+					basePath
+				});
+
+				configs.normalizeSync();
+
+				expect(configs.isFileIgnored(path.join(basePath, 'foo/a.js'))).to.be.true;
+			});
+
+			it('should return false when file is in a directory that is ignored along its files by pattern that ends with `/**` and then unignored along its files by pattern that ends with `/**`', () => {
+				configs = new ConfigArray([
+					{
+						files: ['**/*.js']
+					},
+					{
+						ignores: [
+							'foo/**',
+
+							// both the directory and the files are unignored
+							'!foo/**'
+						]
+					}
+				], {
+					basePath
+				});
+
+				configs.normalizeSync();
+
+				expect(configs.isFileIgnored(path.join(basePath, 'foo/a.js'))).to.be.false;
+			});
+
+			it('should return true when file is ignored by a pattern and there are unignore patterns that target files of a directory with the same name', () => {
+				configs = new ConfigArray([
+					{
+						files: ['**/foo']
+					},
+					{
+						ignores: [
+							'foo',
+							'!foo/*',
+							'!foo/**'
+						]
+					}
+				], {
+					basePath
+				});
+
+				configs.normalizeSync();
+
+				expect(configs.isFileIgnored(path.join(basePath, 'foo'))).to.be.true;
+			});
+
+			it('should return true when file is in a directory that is ignored even if an unignore pattern that ends with `/*` matches the file', () => {
+				configs = new ConfigArray([
+					{
+						files: ['**/*.js']
+					},
+					{
+						ignores: [
+							'foo',
+							'!foo/*'
+						]
+					}
+				], {
+					basePath
+				});
+
+				configs.normalizeSync();
+
+				expect(configs.isFileIgnored(path.join(basePath, 'foo/a.js'))).to.be.true;
+			});
+
+			// https://github.com/eslint/eslint/issues/17964#issuecomment-1879840650
+			it('should return true for all files ignored in a directory tree except for explicitly unignored ones', () => {
+				configs = new ConfigArray([
+					{
+						files: ['**/*.js']
+					},
+					{
+						ignores: [
+
+							// ignore all files and directories
+							'tests/format/**/*',
+
+							// unignore all directories
+							'!tests/format/**/*/',
+
+							// unignore specific files
+							'!tests/format/**/jsfmt.spec.js'
+						]
+					}
+				], {
+					basePath
+				});
+
+				configs.normalizeSync();
+
+				expect(configs.isFileIgnored(path.join(basePath, 'tests/format/foo.js'))).to.be.true;
+				expect(configs.isFileIgnored(path.join(basePath, 'tests/format/jsfmt.spec.js'))).to.be.false;
+				expect(configs.isFileIgnored(path.join(basePath, 'tests/format/subdir/foo.js'))).to.be.true;
+				expect(configs.isFileIgnored(path.join(basePath, 'tests/format/subdir/jsfmt.spec.js'))).to.be.false;
+			});
```

Comment:
These three tests were failing before this change (the files that should have been ignored were considered not ignored).